### PR TITLE
Push down condition check to TiKV for `prewrite` request

### PR DESCRIPTION
**Pull request by zhangjinpeng1987**
### What problem does this PR solve? 
Push down condition check to TiKV for `Insert` request.

### What is changed and how it works?
Push down condition check to TiKV, and the batch_get for insert statement with unique constraint can be removed.

### Check List 

Tests 

 - Unit test

Code changes

 - Has interface methods change

### Related PRs

- [x] https://github.com/pingcap/kvproto/pull/343

- [x] https://github.com/tikv/tikv/pull/4085

These two PRs should be merged first.

### CODE REVIEW

File: store/tikv/2pc.go [hunk 1]
```diff
@@ -97,12 +98,20 @@ func newTwoPhaseCommitter(txn *tikvTxn, connID uint64) (*twoPhaseCommitter, erro
 	mutations := make(map[string]*pb.Mutation)
 	err := txn.us.WalkBuffer(func(k kv.Key, v []byte) error {
 		if len(v) > 0 {
-			mutations[string(k)] = &pb.Mutation{
-				Op:    pb.Op_Put,
-				Key:   k,
-				Value: v,
+			if txn.us.ShouldNotExist(k) {
```
Comment (tiancaiamao) [inline store/tikv/2pc.go]:
```
op := pb.Op_Put
if if txn.us.ShouldNotExist(k) {
    op = pb.Op_Insert
}
mutations[string(k)] = &pb.Mutation{
	Op:    op,
	Key:   k,
	Value: v,
}
```

File: store/mockstore/mocktikv/mvcc_leveldb.go [hunk 2]
```diff
@@ -549,9 +549,35 @@ func prewriteMutation(db *leveldb.DB, batch *leveldb.Batch, mutation *kvrpcpb.Mu
 	if err != nil {
 		return errors.Trace(err)
 	}
-	// Note that it's a write conflict here, even if the value is a rollback one.
-	if ok && dec1.value.commitTS >= startTS {
-		return ErrRetryable("write conflict")
+	if ok {
```
Comment (tiancaiamao) [inline store/mockstore/mocktikv/mvcc_leveldb.go]: What happen when Get() meet Insert Op ?

File: store/mockstore/mocktikv/mvcc_leveldb.go [hunk 3]
```diff
@@ -549,9 +549,35 @@ func prewriteMutation(db *leveldb.DB, batch *leveldb.Batch, mutation *kvrpcpb.Mu
...
+		}
+		// Key should not exist before
+		if mutation.GetOp() == kvrpcpb.Op_Insert {
+ExistCheckingLoop:
+			for {
+				switch dec1.value.valueType {
+				case typePut:
+					return ErrAbort("key already exist")
+				case typeDelete:
+					break ExistCheckingLoop
```
Comment (tiancaiamao) [inline store/mockstore/mocktikv/mvcc_leveldb.go]: We can't break this search until `dec1.value.commitTS < startTS`
```
Insert(ts = curr = 10) -> Delete(ts = 12) -> Put(ts = 9) -> older version -> ...
```

File: store/mockstore/mocktikv/mvcc_leveldb.go [hunk 4]
```diff
@@ -549,9 +549,35 @@ func prewriteMutation(db *leveldb.DB, batch *leveldb.Batch, mutation *kvrpcpb.Mu
...
+				default:
+				}
+
+				dec1 = valueDecoder{
+					expectKey: mutation.Key,
+				}
+				ok, err = dec1.Decode(iter)
+				if err != nil {
+					return errors.Trace(err)
```
Comment (tiancaiamao) [inline store/mockstore/mocktikv/mvcc_leveldb.go]: If there is no previous data, will `Decode` return an error ?

**Comment by tiancaiamao** [comment]
I suggest handle `prewriteMutation` logic like this:

```
func prewriteMutation() {
    if mutation.Op == Op_Insert {
        Get() // Special handle for Op_insert, call the Get() operation
    }
    ....
}
```

This may result in worse performance, but the code is much simpler.
@zhangjinpeng1987

File: store/mockstore/mocktikv/mvcc_leveldb.go [hunk 5]
```diff
@@ -549,16 +567,19 @@ func prewriteMutation(db *leveldb.DB, batch *leveldb.Batch, mutation *kvrpcpb.Mu
 	if err != nil {
 		return errors.Trace(err)
 	}
-	// Note that it's a write conflict here, even if the value is a rollback one.
```
Comment (tiancaiamao) [inline store/mockstore/mocktikv/mvcc_leveldb.go]: Why this comment is removed?

File: store/tikv/2pc.go [hunk 6]
```diff
@@ -97,12 +98,20 @@ func newTwoPhaseCommitter(txn *tikvTxn, connID uint64) (*twoPhaseCommitter, erro
...
+					Key:   k,
+					Value: v,
+				}
+			} else {
+				mutations[string(k)] = &pb.Mutation{
+					Op:    pb.Op_Put,
+					Key:   k,
+					Value: v,
+				}
+				putCnt++
```
Comment (tiancaiamao) [inline store/tikv/2pc.go]: Insert should also increase `putCnt`

File: store/tikv/2pc.go [hunk 7]
```diff
@@ -381,6 +390,18 @@ func (c *twoPhaseCommitter) prewriteSingleBatch(bo *Backoffer, batch batchKeys)
...
 		var locks []*Lock
 		for _, keyErr := range keyErrs {
+			// Check already exists error
+			if alreadyExist := keyErr.GetAlreadyExist(); alreadyExist != nil {
+				key := alreadyExist.GetKey()
+				err := c.txn.us.LookupConditionErr(key)
+				if err == nil {
+					panic(fmt.Sprintf("con:%d, precondition error for key:%s should not be nil", c.connID, key))
+				}
+				log.Errorf("con:%d key: %s already exists", c.connID, key)
```
Comment (tiancaiamao) [inline store/tikv/2pc.go]: Why log error here? `key already exists` is a normal business logic.

File: kv/union_store.go [hunk 8]
```diff
@@ -14,18 +14,17 @@
...
 
 // UnionStore is a store that wraps a snapshot for read and a BufferStore for buffered write.
 // Also, it provides some transaction related utilities.
 type UnionStore interface {
 	MemBuffer
-	// CheckLazyConditionPairs loads all lazy values from store then checks if all values are matched.
-	// Lazy condition pairs should be checked before transaction commit.
-	CheckLazyConditionPairs() error
+	// Returns true if k should not exist before.
+	ShouldNotExist(k Key) bool
```
Comment (tiancaiamao) [inline kv/union_store.go]: This function is only used in test ?

Comment (zhangjinpeng1987) [reply kv/union_store.go]: Also used in 2pc.go.

File: store/tikv/2pc.go (hunk 7 above)
Comment (zhangjinpeng1987) [reply store/tikv/2pc.go]: I think it is better to log duplicated entry error in TiDB.

File: kv/union_store.go (hunk 8 above)
Comment (tiancaiamao) [inline kv/union_store.go]: `ShouldNotExist` and `LookupConditionErr` are nearly the same, should we merge them into one function ?

Comment (zhangjinpeng1987) [reply kv/union_store.go]: Good idea.

Comment (zhangjinpeng1987) [reply kv/union_store.go]: Done

File: store/tikv/2pc.go (hunk 7 above)
Comment (winkyao) [reply store/tikv/2pc.go]: I agree with tiancaiamao, if we log the duplicated error, there will be too many useless logs.

**Comment by zhangjinpeng1987** [comment]
@disksing PTAL

File: kv/union_store.go [hunk 9]
```diff
@@ -201,30 +206,9 @@ func (us *unionStore) markLazyConditionPair(k Key, v []byte, e error) {
...
-		if len(v.value) == 0 {
-			if _, exist := values[k]; exist {
-				return errors.Trace(v.err)
-			}
-		} else {
-			if !bytes.Equal(values[k], v.value) {
-				return errors.Trace(ErrLazyConditionPairsNotMatch)
-			}
-		}
+func (us *unionStore) LookupConditionPair(k Key) *conditionPair {
```
Comment (disksing) [inline kv/union_store.go]: Do we need configuration to make it compatible with old tikv?

File: store/tikv/2pc.go [hunk 10]
```diff
@@ -21,6 +21,7 @@ import (
 	"sync/atomic"
 	"time"
 
+	"fmt"
```
Comment (disksing) [inline store/tikv/2pc.go]: Move it up with other standard packages.

File: kv/union_store.go (hunk 9 above)
Comment (zhangjinpeng1987) [reply kv/union_store.go]: No need.

**Comment by disksing** [comment]
LGTM

**Comment by zhangjinpeng1987** [comment]
@winkyao PTAL again.

**Comment by tiancaiamao** [comment]
LGTM

**Pull request merged**